### PR TITLE
Ensure that mulitpart bodies are always bytes.

### LINKED ISSUE #1760: BigQuery upload_from_file unicode file-like must be opened in binary-mode if it's more than RESUMABLE_UPLOAD_THRESHOLD, otherwise str-mode.
# Steps to reproduce
```python
import sys
from gcloud.bigquery import Client, SchemaField
from gcloud.bigquery.job import CreateDisposition, WriteDisposition

csv_filename = 'sandwiches.csv'

if len(sys.argv) > 1:
    csv_filename = sys.argv[1]

bq = Client()

ds = bq.dataset('test_unicode')
ds.location = 'EU'


def test_unicode_upload(filename, mode):

    if not ds.exists():
        print('Creating dataset: {}'.format(ds.name))
        ds.create()

    fields = [
        SchemaField('name', 'STRING'),
        SchemaField('main_ingredient', 'STRING'),
    ]

    table = ds.table('sandwiches', fields)

    print('Uploading CSV: {}, mode={!r}'.format(csv_filename, mode))
    table.upload_from_file(
        open(filename, mode),
        encoding='UTF-8',
        source_format='CSV',
        write_disposition=WriteDisposition.WRITE_TRUNCATE,
        create_disposition=CreateDisposition.CREATE_IF_NEEDED)

test_unicode_upload(csv_filename, 'r') # Works

test_unicode_upload(csv_filename, 'rb')
# Fails in http.client.HTTPConnection()._send_request
#
# /usr/lib/python3.4/http/client.py in _send_request(self, method, url,
# body, headers)
#    1178         if isinstance(body, str):
#    1179             # RFC 2616 Section 3.7.1 says that text default has a
#    1180             # default charset of iso-8859-1.
# -> 1181             body = body.encode('iso-8859-1')
#    1182         self.endheaders(body)
#
# UnicodeEncodeError: 'latin-1' codec can't encode characters in position
# 649-650: ordinal not in range(256)
```
## `sandwiches.csv`
```csv
name,main_ingredient
Räksmörgås,Räkor
Baguette,Bröd
```

# Expected behavior

When i send in a binary-mode file-like I expect upload_from_file to pass the data through to BigQuery as-is, and that the BigQuery load job will decode it for me using `encoding=`.

# Enviroment

```console
$ python --version
Python 3.4.3+
$ pip freeze | egrep 'httplib2|gcloud'
gcloud==0.13.0
httplib2==0.9.2

```

### PR DESCRIPTION
**Pull request by tseaver**
Loosely based on [@joar's gist](https://gist.github.com/joar/ce5e54c0befd5ff24c312fcd62caab10)

Fixes #1760.

### CODE REVIEW

**Comment by tseaver** [comment]
Note that opening the file passed to `Table.upload_from_file` in text mode now always fails on Py3k (for both small and large files).  We should likely document that requirement.

File: gcloud/streaming/test_transfer.py [hunk 1]
```diff
@@ -1785,6 +1785,18 @@ def test__send_chunk_w_total_size_stream_exhausted(self):
         self.assertEqual(end, SIZE)
 
 
+def _email_chunk_parser():
+    import six
+    if six.PY3:  # pragma: NO COVER  Python3
+        from email.parser import BytesParser  # pylint: disable=E0611
```
Comment (dhermes) [inline gcloud/streaming/test_transfer.py]: @tseaver Please replace with

```python
        # pylint: disable=no-name-in-module
        from email.parser import BytesParser
        # pylint: enable=no-name-in-module
```

logic being that future maintainers (including future us) won't need to Google for what `E0611` is.

File: gcloud/streaming/transfer.py [hunk 2]
```diff
@@ -849,24 +850,25 @@ def _configure_multipart_request(self, http_request):
...
-        # NOTE: We encode the body, but can't use
-        #       `email.message.Message.as_string` because it prepends
-        #       `> ` to `From ` lines.
-        # NOTE: We must use six.StringIO() instead of io.StringIO() since the
-        #       `email` library uses cStringIO in Py2 and io.StringIO in Py3.
-        stream = six.StringIO()
-        generator = email_generator.Generator(stream, mangle_from_=False)
+        # NOTE: generate multipart message as bytes, not text
+        stream = six.BytesIO()
+        if six.PY3:  # pragma: NO COVER  Pythone
```
Comment (dhermes) [inline gcloud/streaming/transfer.py]: s/Pythone/Python 3, right?

**Comment by dhermes** [comment]
RE: Always failing tables, couldn't we just convert to bytes? Then instead of documenting "this will fail", we'd document "if you don't give us bytes, we're going to guess your encoding and we might get it wrong"

**Comment by tseaver** [comment]
Do we want to be in the business of reading potentially huge files in memory and guessing their encoding?  If we want to try checking for 'stream.mode != 'rb'` (assuming they opened it, rather than passing us some other file-like object) we could raise an exception right away, rather than failing with a much uglier traceback later.

File: gcloud/streaming/test_transfer.py (hunk 1 above)
Comment (tseaver) [inline gcloud/streaming/test_transfer.py]: 56e4d0b

File: gcloud/streaming/transfer.py (hunk 2 above)
Comment (tseaver) [inline gcloud/streaming/transfer.py]: efc9ffd

**Comment by tseaver** [comment]
@dhermes  I'm reluctant to get into the mess of transcoding a large input stream opened in text mode.

@jgeewax WDYT?

**Comment by jgeewax** [comment]
I agree that guessing the format is probably a bad idea (because while it might make some people happy, it might make some other people really really mad).

Can we have a super clear exception saying that you gave us a file opened in text mode, which means we don't know the encoding... Can you open in binary mode with the right encoding?

**Comment by dhermes** [comment]
Sorry for the delay. @tseaver I agree with you, sniffing the entire file was ill-conceived.

File: gcloud/streaming/test_transfer.py [hunk 3]
```diff
@@ -1785,6 +1785,20 @@ def test__send_chunk_w_total_size_stream_exhausted(self):
         self.assertEqual(end, SIZE)
 
 
+def _email_chunk_parser():
+    import six
+    if six.PY3:  # pragma: NO COVER  Python3
+        # pylint: disable=no-name-in-module
+        from email.parser import BytesParser
+        # pylint: enable=no-name-in-module
```
Comment (dhermes) [inline gcloud/streaming/test_transfer.py]: @tseaver you can ditch this pylint ignore after #1780?

Comment (tseaver) [inline gcloud/streaming/test_transfer.py]: ceb1c07

**Comment by tseaver** [comment]
@jgeewax 
> Can we have a super clear exception saying ... "You gave us a file opened in text mode, which means we don't know the encoding... Can you open in binary mode with the right encoding?"

IFF the user gives us an actual file object (created via `open`), we can sniff its `mode` attribute.  If they hand us a file-ish object (e.g., an instance of `io.StreamIO`) it likely won't have such an attribute.

**Comment by jgeewax** [comment]
@dhermes And in that case, we just do our best here, right?

**Comment by tseaver** [comment]
@jgeewax 4a8e952 adds a test for a text-mode file, where the exception message says "use open(filename, mode='rb')".

**Comment by tseaver** [comment]
@dhermes, @jgeewax  PTAL: I think I've addressed your concerns.

**Comment by dhermes** [comment]
LGTM

**Pull request merged**